gems/aws-crt/native/src/signable.c

```diff
@@ -0,0 +1,340 @@
+    }
+
+    struct aws_string *name = NULL;
+    struct aws_string *value = NULL;
+
+    name = aws_string_new_from_c_str(signable->allocator, property_name);
+    value = aws_string_new_from_c_str(signable->allocator, property_value);
+
+    if (aws_hash_table_put(&impl->properties, name, value, NULL)) {
+        goto on_error;
+    }
+
+    return AWS_OP_SUCCESS;
+
+on_error:
+
+    aws_string_destroy(name);
+    aws_string_destroy(value);
+
+    return AWS_OP_ERR;
+}
+
+static struct aws_array_list *s_get_or_create_property_list(
+    struct aws_signable *signable,
+    const struct aws_string *list_name) {
+    struct aws_crt_signable_impl *impl = signable->impl;
+    struct aws_hash_element *element = NULL;
+
+    aws_hash_table_find(&impl->property_lists, list_name, &element);
+
+    if (element != NULL) {
+        return element->value;
+    }
+
+    struct aws_array_list *properties = aws_mem_acquire(signable->allocator, sizeof(struct aws_array_list));
+
+    AWS_ZERO_STRUCT(*properties);
+    struct aws_string *name_copy = aws_string_new_from_string(signable->allocator, list_name);
+
+    if (aws_array_list_init_dynamic(
+            properties,
+            signable->allocator,
+            INITIAL_AWS_CRT_SIGNABLE_PROPERTY_LIST_SIZE,
+            sizeof(struct aws_signable_property_list_pair))) {
+        goto on_error;
+    }
+
+    if (aws_hash_table_put(&impl->property_lists, name_copy, properties, NULL)) {
+        goto on_error;
+    }
+
+    return properties;
+
+on_error:
+
+    aws_string_destroy(name_copy);
+    aws_array_list_clean_up(properties);
+    aws_mem_release(signable->allocator, properties);
+
+    return NULL;
+}
+
+int aws_crt_signable_append_property_list(
+    struct aws_signable *signable,
+    const char *list_name,
+    const char *property_name,
+    const char *property_value) {
+
+    if (signable == NULL) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    struct aws_crt_signable_impl *impl = signable->impl;
+    if (impl == NULL) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        ;
+    }
+
+    struct aws_string *list_name_str = NULL;
+    struct aws_string *name = NULL;
+    struct aws_string *value = NULL;
+
+    list_name_str = aws_string_new_from_c_str(signable->allocator, list_name);
+
+    struct aws_array_list *properties = s_get_or_create_property_list(signable, list_name_str);
+    aws_string_destroy(list_name_str);
+
+    name = aws_string_new_from_c_str(signable->allocator, property_name);
+    value = aws_string_new_from_c_str(signable->allocator, property_value);
+
+    struct aws_signable_property_list_pair property;
+    property.name = aws_byte_cursor_from_string(name);
+    property.value = aws_byte_cursor_from_string(value);
+
+    aws_array_list_push_back(&impl->str_buffer, &name);
+    aws_array_list_push_back(&impl->str_buffer, &value);
+
+    aws_string_destroy(name);
+    aws_string_destroy(value);
```

Comment:
why are `name` and `value` being destroyed? Didn't we just push their pointers in the array-list?

Reply:
Good question!  Yes, those shouldn't be getting destroyed there.  Removed that.